File: services/campaign/src/infrastructure/external/kafka_event_publisher.py
import structlog

from ...application.interfaces.event_publisher import EventPublisher, DomainEvent


logger = structlog.get_logger()


class KafkaEventPublisher(EventPublisher):
    """Kafka event publisher implementation"""
    
    def __init__(self, bootstrap_servers: str, client_id: str):
        self.bootstrap_servers = bootstrap_servers
        self.client_id = client_id
        logger.info(
            "Initialized Kafka event publisher",
            servers=bootstrap_servers,
            client_id=client_id
        )
    
    async def publish(self, event: DomainEvent) -> None:
        """Publish domain event to Kafka"""
        # TODO: Implement actual Kafka publishing
        logger.info(
            "Publishing event (mock)",
            event_type=event.event_type,
            event_id=event.event_id,
            aggregate_id=event.aggregate_id
        )
    
    async def publish_batch(self, events: list[DomainEvent]) -> None:
        """Publish multiple events"""
        # TODO: Implement batch publishing
        logger.info("Publishing batch of events (mock)", count=len(events))
        for event in events:
            await self.publish(event)

File: services/content/src/infrastructure/external/KafkaEventPublisher.ts
import { EventPublisher, DomainEvent } from '@application/interfaces/EventPublisher.js';

export class KafkaEventPublisher implements EventPublisher {
  private kafkaProducer: any; // In real implementation, would use Kafka client

  constructor(config: { brokers: string[]; clientId: string }) {
    // Initialize Kafka producer
    console.log('Initializing Kafka producer...', config);
    // this.kafkaProducer = new Kafka(config).producer();
  }

  async publish(event: DomainEvent): Promise<void> {
    console.log(`Publishing event: ${event.eventType}`, {
      eventId: event.eventId,
      aggregateId: event.aggregateId,
      timestamp: event.timestamp,
      payload: event.payload
    });

    // In real implementation:
    // await this.kafkaProducer.send({
    //   topic: this.getTopicForEvent(event.eventType),
    //   messages: [{
    //     key: event.aggregateId,
    //     value: JSON.stringify(event),
    //     headers: {
    //       'event-type': event.eventType,
    //       'event-id': event.eventId,
    //       'timestamp': event.timestamp.toISOString()
    //     }
    //   }]
    // });
  }

  async publishBatch(events: DomainEvent[]): Promise<void> {
    console.log(`Publishing ${events.length} events in batch`);
    
    // In real implementation, would send all events in a single Kafka transaction
    for (const event of events) {
      await this.publish(event);
    }
  }

  private getTopicForEvent(eventType: string): string {
    // Map event types to Kafka topics
    const topicMap: Record<string, string> = {
      'content.created': 'content-events',
      'content.published': 'content-events',
      'content.archived': 'content-events',
      'content.deleted': 'content-events'
    };

    return topicMap[eventType] || 'domain-events';
  }
}